### PR TITLE
p2p: add PoissonNextSend method that returns mockable time

### PR DESCRIPTION
**Pull request by amitiuttarwar**
Introduce a Poisson helper method that wraps the existing method to return `std::chrono::duration` type, which is mockable. 

Needed for https://github.com/bitcoin/bitcoin/pull/16698.

### CODE REVIEW

**Review by maflcko — APPROVED** [review]
Please switch `nNextInvSend` to the new type and use the new function there. Also, a unit test to assert that the two methods return the same value would be nice.

**Comment by amitiuttarwar** [comment]
@MarcoFalke updated `nNextInvSend`. Not sure how to unit test since the result of the poisson method would be different for two separate calls. LMK if you want deeper changes with chrono types, happy to update.

**Comment by MarcoFalke** [comment]
You can set `g_mock_deterministic_tests = true;` for just your test case, then set it to `false` again at the end.

File: src/net.h [hunk 1]
```diff
@@ -943,11 +943,13 @@ class CNode
...
 };
 
-
-
-
-
 /** Return a timestamp in the future (in microseconds) for exponentially distributed events. */
 int64_t PoissonNextSend(int64_t now, int average_interval_seconds);
 
+/** Wrapper to return mockable type */
```
[the remark this answers is not in the document]
Comment (amitiuttarwar) [reply src/net.h]: I'd rather have the comment draw attention to the fact that its just calling through to the other Poisson method. Esp with the function signature & comment right there.

**Comment by amitiuttarwar** [comment]
@MarcoFalke added tests. Ready for review.

**Comment by MarcoFalke** [comment]
ACK d504d70e602bf30aad18e1e5677daa3961757bb7

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK d504d70e602bf30aad18e1e5677daa3961757bb7
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUgYJAv+Nkm4e1Ac/wfPzNKwyAfkC7oRMCGqY1IqPm0vCugazxm+02u7JFotM4rI
j75ONrqy/LC7TAk7ELIWUfJv0lquBLGpkz5ZuYAcfAH8CjSj9NEt1bfm+XRmmXFe
jxLvL068KeDbz/6i2vwwQW3557/0A5p1z8zmTh66IXF7lQU/yjpSAOK5BXXxoVs8
sMqj2o1vZ7TMPZlGTxZxMQB49DUxdRi6TYMKssS7lJqrVbtTxHZcTwzdfOWWVg6e
epAow+pvO1GRV0yYjMzhAXH8HM3500GI3b66gVvvDpIbQK9+pvYDFuVadSHDYvpK
cbpbu5uXhaskeI5OOiv03L3v7lHrRQi/y01w03mD3SgGW66nE2YkN8L6+I5mk6Li
LovIQIvTr6x62xjt5YqYGZW+35RRYdIg1tueNKqmCSAYRGo7kBjUnDyCA3Y8/JGB
jv+afA/RUT2U+8lHYyZkV6rSr9HRuq16DKutyYGiz4VWeyGC0OtUCInf0fvnJkY8
380rfFsr
=RKgu
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `e887cddc7cfb3c4288a10e649645fa1687a1de16f54a8b1819200b2eaa1c86a8  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108e887cddc7cfb3c4288a10e649645fa1687a1de16f54a8b1819200b2eaa1c86a8f0101db986a2e758261b39e489380691715808fff01024ce3478bead8c5f8d02021a0c14008e08f1045dbc9113f0084ce9e5fe5aabacda0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff0103870883ab50e09b280c2ea1aaf687b9c08f1045dbc9113f0085571c02bc44b1f2e0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6df010e91ecfe3f02a7acae1be5c1a157ec56d08f1204aa13daef47b8e93242684a6c0c1ec7e07ea029de567d360fc21657d13dcd90408f1045dbc9113f008efec617d464ff5150083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6d)

</details>

**Review by ariard — COMMENTED** [review]
Can you explain in commit message why you switch for mockable time for outbound peers announcement ? I mean how it's going to be used in #16698

File: src/net_processing.cpp [hunk 2]
```diff
@@ -3548,6 +3548,8 @@ bool PeerLogicValidation::SendMessages(CNode* pto)
 
         // Address refresh broadcast
         int64_t nNow = GetTimeMicros();
+        auto current_time = GetTime<std::chrono::microseconds>();
```
[the remark this answers is not in the document]
Comment (amitiuttarwar) [reply src/net_processing.cpp]: tried to introduce scoping with `{ }` around the inventory code but found that `fFetch` is assigned between the two definitions of `current_time` and accessed after. is there a way to scope independently?

File: src/test/net_tests.cpp [hunk 3]
```diff
@@ -303,5 +303,22 @@ BOOST_AUTO_TEST_CASE(LocalAddress_BasicLifecycle)
...
 
+BOOST_AUTO_TEST_CASE(PoissonNextSend)
+{
+    g_mock_deterministic_tests = true;
+
+    int64_t now = 5000;
+    int average_interval_seconds = 600;
+
+    std::chrono::microseconds now_chrono = std::chrono::microseconds{now};
+    std::chrono::seconds interval_chrono = std::chrono::seconds{average_interval_seconds};
```
Comment (ajtowns) [inline src/test/net_tests.cpp]: `std::chrono::microseconds now_chrono{now};` likewise, or could just say `::PoissonNextSend(std::chrono::microseconds{now}, std::chrono::seconds{average_interval_seconds})`

**Review by ajtowns — APPROVED** [review]
ACK d504d70e602bf30aad18e1e5677daa3961757bb7 nits only

File: src/net_processing.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (ariard) [reply src/net_processing.cpp]: This compile : https://github.com/ariard/bitcoin/commit/1c1765e636db2c9484d10712596754dd774ed7bd, or are you describing a behavior change related with `fFetch` than I miss ?

**Comment by ariard** [comment]
ACK d504d70, checked directly on parent PR the rational of intended changes

**Comment by MarcoFalke** [comment]
@amitiuttarwar This has three ACKs, so is ready to go. Do you want to address the nits or leave them for later?

**Comment by naumenkogs** [comment]
ACK d504d70
willing to do quick re-ACK if those tiny nits are resolved :)

**Comment by ajtowns** [comment]
ACK 1a8f0d5a74d5cc0000456932babf35301f5c1686

**Comment by MarcoFalke** [comment]
re-ACK 1a8f0d5a74d5cc0000456932babf35301f5c1686

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

re-ACK 1a8f0d5a74d5cc0000456932babf35301f5c1686
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUjapAv+MgqGwdYAgJg/hzj7rhv2Fp+bXSWfAoF4Sgqeo7yZbQLfupqRev/uf3hP
DgC6rpyqIM2X8JuijUVxyOs2Z+bU8pOZMOpzVOGpZPHcgmBP1lVT+BqoV3iVgwD2
6dihYwbFOZIZz4jRyAn8Ca9Aoo4jRXY82kPKhhLfo4lRIcp19v0KzccyLeE3rwzh
Do7I5M0PISDid6t2xv57+8R+W5IZy5VChShEx3aHkGLVWR5Qwa01WnoPnPV5uQJO
amkT+i0i5TIRyyN8vSIJWLHfa7BJmG3cRAwekoq0hEeaFw3513hMFPtq+ekfOKb+
DdNFVFYPk5tkTfnYDHcjzxob2phFtEDpPxII8miRu1pFFhkSmcEUswEQZMCykcqi
CBddH8/oGs0G7yPjPtJ6GrmITsJHRwxhlxTyXarx3dm7F78AomLvDYRWUncSgGrh
X7k7PP2ZYTIZIe6YBiPQ1Y6zf73qdw52bsMtXhD5Fj9v/KTZMcW93IghKppmx6Hl
vnB9QRAE
=kQL2
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `c18dd002e27bda4c5087dbc652b61c6e0ece0bbd73a3416548d7cf0624aca3fb  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108c18dd002e27bda4c5087dbc652b61c6e0ece0bbd73a3416548d7cf0624aca3fbf010d014d9594267e5a6a64858febb95a67908fff0100e59517bdb89f3f1c76e8cf58b8be17808f020656b78849c34856f5c90264f6eebda952a2bb9a2add953d6a10a700c6f39808208f1045dc1949ef0086fcb1d5d07f0c60e0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff0102779e3c4d38258a126ea5148e759653f08f1045dc1949df008e3099f02c067cb450083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff0107b43e3b060dd822bd14f225babaadd1d08f1045dc1949df008f50a1ee6dbf0df5a0083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6df010d3cc636264fa3963665dfb30cebfdffe08f1045dc1949df00888477de86c5f78020083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

**Comment by naumenkogs** [comment]
Is it a good idea to have both nNow and current_time? This might lead to counter-intuitive consequences, specifically in the test cases (I don't have the exact example right now).

**Comment by naumenkogs** [comment]
>tests based on what functionality executes based on clock time vs needing to update mocktime. Is that what you're alluding to?

Yes. Also, potentially future code in net_processing.
At the same time, I guess it blocks you from proceeding with rebroadcast (even in terms of better testing) **today.**

ACK 1a8f0d5, and let's merge it and come back to it later.

**Pull request merged**